### PR TITLE
Verify: no ECS suffix for Kibana object filenames

### PR DESCRIPTION
**Pull request by mtojek**
Issue: https://github.com/elastic/package-spec/issues/71

Validate Kibana object filenames in terms of forbidden ECS suffix.

### CODE REVIEW

File: code/go/internal/validator/folder_item_spec.go [hunk 1]
```diff
@@ -14,15 +14,16 @@ import (
...
-	Name             string `yaml:"name"`
-	Pattern          string `yaml:"pattern"`
-	Required         bool   `yaml:"required"`
-	Ref              string `yaml:"$ref"`
-	Visibility       string `yaml:"visibility" default:"public"`
-	commonSpec       `yaml:",inline"`
+	Description       string   `yaml:"description"`
+	ItemType          string   `yaml:"type"`
+	ContentMediaType  string   `yaml:"contentMediaType"`
+	ForbiddenPatterns []string `yaml:"forbiddenPatterns"`
```
Comment (ycombinator) [inline code/go/internal/validator/folder_item_spec.go:20]: If we can _somehow_ make the regular expression work (per our conversation off-PR), we wouldn't need this field, right?

For now, could you move it to below the `Pattern` field in the struct since they are related.

File: code/go/internal/validator/test/packages/wrong_kibana_filename/manifest.yml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 format_version: 1.0.4
-name: good
-title: GOOD
-description: This package is good.
+name: bad
+title: BAD
+description: This package is bad.
```
Comment (ycombinator) [inline code/go/internal/validator/test/packages/wrong_kibana_filename/manifest.yml:4]: Thanks for fixing this, I missed it in an earlier review, sorry!

File: code/go/pkg/validator/validator_test.go [hunk 3]
```diff
@@ -59,19 +59,21 @@ func TestValidateItemNotAllowed(t *testing.T) {
...
 			},
 			"kibana/search": []string{
 				"691240b5-7ec9-4fd7-8750-4ef97944f960.json",
-				"bad-search.json",
+				"missed-search.json",
 			},
 			"kibana/visualization": []string{
 				"defa1bcc-1ab6-4069-adec-8c997b069a5e.json",
-				"bad-visualization.json",
+				"missed-visualization.json",
```
Comment (ycombinator) [inline code/go/pkg/validator/validator_test.go:76]: Why `missed-`?

File: versions/1/kibana/spec.yml [hunk 4]
```diff
@@ -10,6 +10,8 @@
         type: file
         contentMediaType: "application/json"
         pattern: '{PACKAGE_NAME}-.+\.json'
+        forbiddenPatterns:
+          - '.+-(ecs|ECS)\.json$' # ECS suffix is forbidden
```
Comment (ycombinator) [inline versions/1/kibana/spec.yml:14]: ❤️  the comments here.

File: code/go/internal/validator/folder_item_spec.go (hunk 1 above)
Comment (mtojek) [reply code/go/internal/validator/folder_item_spec.go:20]: Yes, we wouldn't need it.

Apart from that, it's MUCH simpler and easier to read, having two independent patterns. See: https://github.com/elastic/elastic-package/pull/154#discussion_r515173374

File: code/go/pkg/validator/validator_test.go (hunk 3 above)
Comment (mtojek) [reply code/go/pkg/validator/validator_test.go]: I renamed to `not-bad` as it's different to `bad`.

Then I discoved another issue with regular expressions - the spec doesn't care about `^$`, so I had to consider them in the validator. It failed datastream regexp, so I adjusted it.

File: code/go/pkg/validator/validator_test.go [hunk 5]
```diff
@@ -59,19 +59,21 @@ func TestValidateItemNotAllowed(t *testing.T) {
...
 				"b7e55b73-97cc-44fd-8555-d01b7e13e70d.json",
-				"bad-dashboard.json",
+				"bad-ecs.json",
+				"bad-foobar-ecs.json",
+				"bad-Foobaz-ECS.json",
 			},
 			"kibana/map": []string{
 				"06149856-cbc1-4988-a93a-815915c4408e.json",
-				"bad-map.json",
+				"not-bad-map.json",
```
Comment (ycombinator) [inline code/go/pkg/validator/validator_test.go:68]: Sorry, I'm a little confused about this name change. What was the problem if you left it as `bad-map.json`?

Comment (mtojek) [reply code/go/pkg/validator/validator_test.go:68]: If you take a look at `test/package`, the name `good` is restricted to the package which is 100% aligned to the schema.   This package here is not good - `bad` :) which means that `bad-map.json` is a valid Kibana file, but I need a one with wrong name here.

Comment (ycombinator) [reply code/go/pkg/validator/validator_test.go:68]: Ah, got it, thanks for explaining. `not-bad` is quite a twisted way of saying `good` so how about we just give these files UUIDs as names? Anyway I don't see any difference between this file and the one above it (`06149856-cbc1-4988-a93a-815915c4408e.json`) - they are both files with empty JSON objects in them and neither of them have a `-ecs` suffix in the name. Or am I missing something again? 😄

File: versions/1/data_stream/spec.yml [hunk 6]
```diff
@@ -3,7 +3,7 @@ spec:
   contents:
   - description: Folder containing a single data stream definition
     type: folder
-    pattern: '[a-z0-9]([a-z0-9_]+[a-z0-9])?'
+    pattern: '([a-z0-9][a-z0-9]|[a-z0-9][a-z0-9_][a-z0-9_]+)'
```
Comment (ycombinator) [inline versions/1/data_stream/spec.yml:6]: Why was this change necessary?

File: code/go/pkg/validator/validator_test.go (hunk 5 above)
Comment (mtojek) [reply code/go/pkg/validator/validator_test.go:68]: The intention was:
* `06149856-cbc1-4988-a93a-815915c4408e.json` is a wrong file without package name prefixed.
* `not-bad-map.json` is a file with a different package name.

I suppose I can fallback to the first option only (keep the `06149856-cbc1-4988-a93a-815915c4408e.json` file only).

File: versions/1/data_stream/spec.yml (hunk 6 above)
Comment (mtojek) [reply versions/1/data_stream/spec.yml:6]: `pe` datastream was not valid.

Comment (mtojek) [reply versions/1/data_stream/spec.yml]: BTW I changed it now to `'([a-z0-9][a-z0-9]|[a-z0-9][a-z0-9_]+[a-z0-9])'

File: code/go/pkg/validator/validator_test.go (hunk 5 above)
Comment (ycombinator) [reply code/go/pkg/validator/validator_test.go:68]: I see. Thanks for clarifying. I think the issue is that it's not obvious from the filenames what purpose they are meant to serve in the tests. I would suggest keeping two files, one to test each case that you explained. But:
- maybe instead of calling it `not-bad-map`, call it something like `another-package-map`, only because `not-bad` is confusing ("why didn't they just call it `good`", someone looking at this file might ask), and
- maybe flesh out each of the files just enough to get a description in there, where we can see what purpose that file is serving in tests.

File: versions/1/data_stream/spec.yml (hunk 6 above)
Comment (ycombinator) [reply versions/1/data_stream/spec.yml]: Understood. BTW, you can change the first option in the regex (the left side of the `|`) to `[a-z0-9]{2}`.

File: code/go/pkg/validator/validator_test.go (hunk 5 above)
Comment (mtojek) [reply code/go/pkg/validator/validator_test.go:68]: > maybe instead of calling it not-bad-map, call it something like another-package-map, only because not-bad is confusing ("why didn't they just call it good", someone looking at this file might ask), and

`another-package-map.json` LGTM , I will adjust the code.

> maybe flesh out each of the files just enough to get a description in there, where we can see what purpose that file is serving in tests.

I can add two properties, ID and description, to add more context to the test case. I wouldn't pollute it with a real Kibana object with all properties as I believe it's not necessary them.

Comment (ycombinator) [reply code/go/pkg/validator/validator_test.go:68]: > I can add two properties, ID and description, to add more context to the test case. I wouldn't pollute it with a real Kibana object with all properties as I believe it's not necessary them.

👍  I think it might be good to follow a real Kibana object's schema though. That way if we ever start validating Kibana object structures, we don't suddenly start failing this test. But ++ to only populating the minimum number of properties needed.

File: versions/1/data_stream/spec.yml (hunk 6 above)
Comment (mtojek) [reply versions/1/data_stream/spec.yml]: Nice, I simplified the regular expression.

File: code/go/pkg/validator/validator_test.go (hunk 5 above)
Comment (mtojek) [reply code/go/pkg/validator/validator_test.go]: Done

**Review by ycombinator — APPROVED** [review]
LGTM. Thanks for the clarifying changes.
